pkg/k8s/watchers/cilium_cidr_group.go

```diff
@@ -0,0 +1,281 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package watchers
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	cilium_v2_alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	"github.com/cilium/cilium/pkg/k8s/watchers/resources"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/policy/api"
+)
+
+func (k *K8sWatcher) onUpsertCIDRGroup(
+	cidrGroup *cilium_v2_alpha1.CiliumCIDRGroup,
+	cidrGroupCache map[string]*cilium_v2_alpha1.CiliumCIDRGroup,
+	cnpCache map[resource.Key]*types.SlimCNP,
+	cs client.Clientset,
+	apiGroup, metricLabel string,
+) error {
+	var (
+		errs []error
+
+		equal  bool
+		action string
+	)
+
+	// wrap k.K8sEventReceived call into a naked func() to capture equal in the closure
+	defer func() {
+		k.K8sEventReceived(apiGroup, metricLabel, action, true, equal)
+	}()
+
+	oldCidrGroup, ok := cidrGroupCache[cidrGroup.Name]
+	if ok && oldCidrGroup.Spec.DeepEqual(&cidrGroup.Spec) {
+		equal = true
+		return nil
+	} else if ok {
+		action = resources.MetricUpdate
+	} else {
+		action = resources.MetricCreate
+	}
+
+	cidrGroupCpy := cidrGroup.DeepCopy()
+	cidrGroupCache[cidrGroup.Name] = cidrGroupCpy
+
+	for key, cnp := range cnpCache {
+		if !hasCIDRGroupRef(cnp, cidrGroup.Name) || cnp.RequiresDerivative() {
+			continue
+		}
+
+		log.WithFields(logrus.Fields{
+			logfields.CiliumNetworkPolicyName: cnp.Name,
+			logfields.K8sAPIVersion:           cnp.APIVersion,
+			logfields.K8sNamespace:            cnp.Namespace,
+			logfields.CIDRGroupRef:            cidrGroup.Name,
+		}).Info("Referenced CiliumCIDRGroup updated, recalculating CiliumNetworkPolicy rules")
+
+		initialRecvTime := time.Now()
+
+		// We need to deepcopy this structure because we are writing
+		// fields.
+		// See https://github.com/cilium/cilium/blob/27fee207f5422c95479422162e9ea0d2f2b6c770/pkg/policy/api/ingress.go#L112-L134
+		cnpCpy := cnp.DeepCopy()
+
+		translationStart := time.Now()
+		translatedCNP := resolveCIDRGroupRef(cnpCpy, cidrGroupCache)
+		metrics.CIDRGroupTranslationTimeStats.WithLabelValues(cnpCpy.Namespace, cnpCpy.Name).
+			Observe(time.Since(translationStart).Seconds())
+
+		err := k.updateCiliumNetworkPolicyV2(cs, cnpCpy, translatedCNP, initialRecvTime)
+		if err == nil {
+			cnpCache[key] = cnpCpy
+		}
+
+		errs = append(errs, err)
+	}
+
+	err := errors.Join(errs...)
+
+	k.K8sEventProcessed(metricLabel, action, err == nil)
+
+	return err
+}
+
+func (k *K8sWatcher) onDeleteCIDRGroup(
```

Comment:
nit: I see a few minor differences between `onDeleteCIDRGroup()` and `onUpsertCIDRGroup()`, but in the end they both translate into calls to update the network policy for each CNP that selects this CIDRGroup. Is there any reason to have two different functions or could they be the same handler?

Reply:
Apart from the differences in cache management and metrics reporting the code is equal, I moved it in a separate function to be reused. :+1: